distro/src/avail/Avail.avail/Foundation.avail/Late Math.avail

```diff
@@ -846,3 +868,247 @@ Public stable method "_(decimal)" is
+[
+	base : whole number,
+	exponent : whole number,
+	modulus : natural number
+|
+	moddedBase : whole number := base mod modulus;
+	exponentRemaining : integer := exponent;
+	result : whole number := 1;
+	Until exponentRemaining = 0 do
+	[
+		If exponentRemaining bit∧ 1 = 1 then
+		[
+			result := (result × moddedBase) mod modulus;
+		];
+		exponentRemaining := (exponentRemaining >> 1);
+		moddedBase := (moddedBase × moddedBase) mod modulus;
+	];
+	result
+];
+
+/**
+ * Restrict modular exponentiation to the combined restrictions upon the unfused
+ * exponentiation and modulus methods.
+ *
+ * @method "_^_mod_"
+ * @restricts "whole number's type"
+ * @restricts "whole number's type"
+ * @restricts "natural number's type"
+ */
+Semantic restriction "_^_mod_" is
+[
+	base : whole number's type,
+	exponent : whole number's type,
+	modulus : natural number's type
+|
+	(base ^ exponent) mod modulus
+];
+
+Grammatical restriction "_^_" is <∅, {"_mod_"}>;
+Grammatical restriction "_mod_" is <{"_^_"}, ∅>;
+
+/**
+ * Determine if an integer is a "strong probable prime" to a given base (said to
+ * be "b-SPRP" for short). Prime numbers will always return true for this test
+ * but some composites will as well, known as "strong pseudoprimes".
+ *
+ * Evaluating whether a number is a strong probable prime is a relatively fast 
+ * and useful component of several probabilistic techniques for determining
+ * primality. Even checking against a random base in [2, n-2] returns
+ * false for ~75% of composites -- this particular process is known as the
+ * Miller-Rabin test.
+ *
+ * @method "_is a strong probable prime to base_"
+ * @param "n" "natural number"
+ * @param "base" "natural number"
+ * @returns "boolean"
+ *
+ * @category "Mathematics" "Integers" "Types"
+ */
+Public stable method "_is a strong probable prime to base_" is
+[
+	n : natural number,
+	base : natural number
+|
+	$body : boolean;
+	If n = 1 ∨ n = base ∨ (n ≠ 2 ∧ n is even) then [Exit body with false];
+	If base = 1 then [Exit body with true];
+	d : whole number := n - 1;
+	s : whole number := 0;
+	While d is even do
+	[
+		s := s + 1;
+		d := d >> 1;
+	];
+
+	test : whole number := base ^ d mod n;
+	if test ∈ {1, n-1} then [true]
+	else
+	[
+		first index of 1 to (s-1) where
+		[
+			i : natural number
+		|
+			test := test × test mod n;
+			test = n - 1
+		] > 0
+	]
+];
+
+smallPrimes ::= {
+	2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
+	73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127
+};
+
+/**
+ * Determine if a {@type "[0..130]" small integer} is prime. This method performs
+ * a lookup against a list of the first 31 primes.
+ *
+ * @method "_is prime"
+ * @param "x" "whole number"
+ * @returns "boolean"
+ *          {@method "true"} if {@param "x"} is prime, {@method "false"}
+ *          otherwise.
+ * @category "Mathematics" "Integers" "Numbers"
+ */
+Public method "_is prime" is
+[
+	x : [0..130]
+|
+	x∈smallPrimes
+];
```

Comment:
Agreed, but the general case method should be an actual implementation of an exact primality test.  Defining the abstract method and overriding it for [131..∞) seems risky, by repeating the threshold (+1 in this case) in two places.  Divisor testing in a loop should suffice, although a sieve is technically better performance asymptotically.  It's questionable whether it's worthwhile to statistically check for primality first, and answer false if it was determined to be definitely composite (and falling back on the sieve otherwise).  I wouldn't bother with the pre-check.

Reply:
Ok I'm happy to put a base implementation in if that's preferred. I shied away from it because the decent algorithms at that scale are complex, but I understand getting the type coverage might be more important than only accepting code with good performance.

> It's questionable whether it's worthwhile to statistically check for primality first

Funny, I see so much advice on the internet that starts with "first throw Miller-Rabin at it" but it's hard to judge whether those posts come from any point of experience. In any case I already have the statistical test, so I'll try it both with and without.